sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala

```diff
@@ -68,50 +69,91 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
       cteDef.child.exists(_.expressions.exists(_.isInstanceOf[OuterReference]))
   }
 
+  /**
+   * Accumulates all the CTEs from a plan into a special map.
+   *
+   * @param plan The plan to collect the CTEs from
+   * @param cteMap A mutable map that accumulates the CTEs and their reference information by CTE
+   *               ids. The value of the map is tuple whose elements are:
+   *               - The CTE definition
+   *               - The number of incoming references to the CTE. This includes references from
+   *                 outer CTEs and regular places.
```

Comment:
```suggestion
   *                 inner CTEs and regular places.
```

Reply:
I actually wanted to write `other CTEs` and not inner/outer. E.g. in
```
WITH (
  cte1 AS (SELECT 1),
  cte2 AS (SELECT * FROM cte1)
)
SELECT * FROM cte1 JOIN cte2 ON ...
```
the reference count of cte1 is 2. 1 is from an "other" CTE (but not "inner"/"outer"), 1 is from a "regular place".

Reply:
fixed in https://github.com/apache/spark/pull/40856/commits/806c2de2b7db9b960c591a849dc43be7f65b468b

Reply:
@cloud-fan, please let me know if anything else is needed here.